outbox-kafka-spring/src/main/java/one/tomorrow/transactionaloutbox/service/OutboxProcessor.java

```diff
@@ -41,167 +41,172 @@
+        this.producerFactory = producerFactory;
+        producer = producerFactory.createKafkaProducer();
+
+        executor = Executors.newSingleThreadScheduledExecutor();
+
+        tryLockAcquisition();
+    }
+
+    private void scheduleProcessing() {
+        if (executor.isShutdown())
+            logger.info("Not scheduling processing for lockOwnerId {} (executor is shutdown)", lockOwnerId);
+        else
+            schedule = executor.schedule(this::processOutboxWithLock, processingInterval.toMillis(), MILLISECONDS);
+    }
+
+    private void scheduleTryLockAcquisition() {
+        if (executor.isShutdown())
+            logger.info("Not scheduling acquisition of outbox lock for lockOwnerId {} (executor is shutdown)", lockOwnerId);
+        else
+            schedule = executor.schedule(this::tryLockAcquisition, lockService.getLockTimeout().toMillis(), MILLISECONDS);
+    }
+
+    @PreDestroy
+    public void close() {
+        logger.info("Stopping OutboxProcessor.");
+        if (schedule != null)
+            schedule.cancel(false);
+        executor.shutdown();
+        producer.close();
+        if (active)
+            lockService.releaseLock(lockOwnerId);
+    }
+
+    private void tryLockAcquisition() {
+        try {
+            boolean originalActive = active;
+            logger.debug("{} trying to acquire outbox lock", lockOwnerId);
+            active = lockService.acquireOrRefreshLock(lockOwnerId);
+            lastLockAckquisitionAttempt = now();
+            if (active) {
+                if (originalActive)
+                    logger.debug("{} acquired outbox lock, starting to process outbox", lockOwnerId);
+                else
+                    logger.info("{} acquired outbox lock, starting to process outbox", lockOwnerId);
+
+                processOutboxWithLock();
+            }
+            else
+                scheduleTryLockAcquisition();
+        } catch (Exception e) {
+            logger.warn("Failed trying lock acquisition or processing the outbox, trying again in {}", lockService.getLockTimeout(), e);
+            scheduleTryLockAcquisition();
+        }
+    }
+
+    private void processOutboxWithLock() {
+        if (!active)
+            throw new IllegalStateException("processOutbox must only be run when in active state");
+
+        if (now().isAfter(lastLockAckquisitionAttempt.plus(lockService.getLockTimeout().dividedBy(2)))) {
+            tryLockAcquisition();
+            return;
+        }
+
+        boolean couldRunWithLock = false;
+        try {
+            couldRunWithLock = lockService.runWithLock(lockOwnerId, () -> {
+                try {
+                    processOutbox();
+                } catch (Throwable e) {
+                    logger.warn("Recreating producer, due to failure while processing outbox.", e);
+                    producer.close();
+                    producer = producerFactory.createKafkaProducer();
+                }
+            });
```

Comment:
I don't like having a try-catch in a try-catch. Maybe you could try to exclude this into a sperate method named `tryProcessOutbox()`. WDYT?

Reply:
Done